src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs

```diff
@@ -303,42 +321,56 @@ public void Reset()
             _continuationState = null;
         }
 
+        /// <summary>Completes with a successful result.</summary>
+        /// <param name=""result"">The result.</param>
         public void SetResult(TResult result)
         {
             _result = result;
             SignalCompletion();
         }
 
+        /// <summary>Complets with an error.</summary>
+        /// <param name=""error""></param>
         public void SetException(Exception error)
         {
             _error = ExceptionDispatchInfo.Capture(error);
             SignalCompletion();
         }
 
+        /// <summary>Gets the operation version.</summary>
         public short Version => _version;
 
+        /// <summary>Gets the status of the operation.</summary>
+        /// <param name=""token"">Opaque value that was provided to the <see cref=""ValueTask""/>'s constructor.</param>
         public ValueTaskSourceStatus GetStatus(short token)
         {
             ValidateToken(token);
             return
-                !_completed ? ValueTaskSourceStatus.Pending :
+                _continuation == null || !_completed ? ValueTaskSourceStatus.Pending :
```

Comment:
@stephentoub FWIW, this seems the only actual code change when updating `ManualResetValueTaskSourceCore` source code, but I'm not seeing how it could relate to the symptom (ie. the completion sentinel being invoked, see stacktrace below).
Although almost all the [failures](https://runfo.azurewebsites.net/search/tests/?q=+definition%3Aroslyn-ci++result%3AFailed++started%3A%7E5+message%3A%22Operation+is+not+valid+due+to+the+current+state+of+the+object.%22) occur on Linux, there's one on Windows too (so not exclusive).

```
    at System.Threading.Tasks.Sources.ManualResetValueTaskSourceCoreShared.ThrowInvalidOperationException()
    at System.Threading.Tasks.Sources.ManualResetValueTaskSourceCoreShared.CompletionSentinel(Object _)
    at System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore`1.InvokeContinuation()
    at System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore`1.SignalCompletion()
    at System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore`1.SetResult(TResult result)
    at C.<Main>g__local|0_0()
    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
    at System.Runtime.CompilerServices.AsyncIteratorMethodBuilder.MoveNext[TStateMachine](TStateMachine& stateMachine)
    at C.<Main>g__local|0_0()
    at C.Main()
    at C.Main()
    at C.<Main>()
```

Reply:
That change came from https://github.com/dotnet/coreclr/pull/22711 which was addressing a race condition.  So... it might be relevant.